### PR TITLE
KAFKA-9929: Support reverse iterator on KeyValueStore

### PR DESCRIPTION
**Pull request by jeqo**
Implements [KIP-617](https://cwiki.apache.org/confluence/display/KAFKA/KIP-617%3A+Allow+Kafka+Streams+State+Stores+to+be+iterated+backwards) on KeyValueStore.

Testing strategy: extend existing tests to validate reverse operations are supported.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jeqo** [comment]
cc @ableegoldman 

This PR is ready for review, covering related feedback from #8976

File: streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java:67]: Seems a bit tricky to say that _to_ is the variable where iteration starts _from_ 😉  But I can see it both ways, so being clear in the javadocs is good enough for me

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java]: Should be private

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java]: nit: Use braces & separate lines

File: streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java]: nit: braces + separate lines

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java]: nit: `RawToKey` --> `RawLastKey`

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java]: Also it should be `from` for the reverse case, right?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java]: nit: braces & separate lines

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java]: braces & separate lines

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbIterator.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbIterator.java]: separate lines 🙂

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java]: braces & separate lines here and below

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java:510]: Can we add tests for some other invalid range cases? For example with both bounds positive but from > to

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java:332]: Can we add some tests that verify the actual contents + order of the reverse range?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java]: Use `assertThrows` -- we've been (slowly) migrating away from `@Test(expected)` in the Streams tests

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java]: use `assertThrows` here as well

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java]: I think the best we can do is just make sure that order is correct within a store. ie if `a`, `m` are all in `stubOneUnderling` then make sure the reverse range returns `m` before `a`.

I also think it would be fine to just make sure all the expected values are returned without checking the order, since there are other tests to verify that the order within a store is correct

**Review by ableegoldman — COMMENTED** [review]
Thanks for splitting this up! Mostly minor comments on the KVStore PR, some general notes about the Streams style conventions:
1. always use braces for `if`/`else`/etc and put each condition on its own line
2. We're trying to move away from using `@Test(expected)` in the tests in favor of `assertThrows`. A lot of the tests haven't been migrated over yet, but just a heads up

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java
Comment (jeqo) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBRangeIterator.java]: right. I've moved this into each condition to set a correct exception message.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java:510]: ack. added.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java:332]: ack. added.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java]: migrated the whole class. will apply this to the other PRs.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java]: ack. make sense.

**Comment by jeqo** [comment]
@ableegoldman this should be ready for another review :)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java]: This class is still unused so maybe we should just take it out. Someone's working on a prefix seek KIP at the moment so it'll be replaced soon anyway

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java]: We need to branch on `reverse` here too, right?

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java]: Kind of unrelated, but WDYT about renaming `RocksDBDualCFIterator` to `RocksDBDualCFAllIterator` or something on the side? I feel like these iterators could be cleaned up a bit in general to be more understandable -- for example, it's weird that we do the `iterator#seek`-ing in the actual `all()` method but for range queries we do the seeking inside the iterator constructor. 
Just thinking out loud though, we can do some followup refactoring once this is merged

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java]: If `compare(noTimestamp, withTimestamp)` <= 0 then `withTimestamp >= noTimestamp`, so for the reverse case we would actually want to return `withTimestamp` next here (and vice versa in the else block)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbIterator.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbIterator.java]: Instead of storing and checking the `reverse` flag on every iteration, can we define something like
```
java.util.function.Consumer<RocksIterator> advanceIterator = reverse ? RocksIterator::prev : RocksIterator::next;
```
so then we can just blindly call `advanceIterator.accept(iter)`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java:324]: Awesome, thank you for cleaning up the whole class 🙏

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java:191]: I think we should add some tests to `RocksDBTimestampedStoreTest`. I thought it would extend `AbstractKeyValueStoreTest` and thus benefit from everything you added here, but doesn't seem to be the case :/ 
Personally I found the `RocksDBDualCFIterator` logic a bit difficult to follow even before the reverse iteration,  so it would be nice to have some tests specifically covering reverse iterators over multi-column-family timestamped stores

**Review by ableegoldman — COMMENTED** [review]
Did a quick survey of everywhere that implements `range()` and `all()` to see if there might be anywhere else we need to add `reverseRange` and `reverseAll`. These guys seem to be missing their reverse counterparts: 
 - `KeyValueStoreReadWriteDecorator`
- `KeyValueToTimestampedKeyValueByteStoreAdapter`
- `Segment` (this one is not strictly necessary until the WindowStore PR though)

The fact that you've had to add these to about 25 different classes already really highlights the mess of the store hierarchy 😬

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java:191]: Just realized that, I also thought that path was tested. Good catch!

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java
Comment (jeqo) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java]: Agree. I will continue the current approoach and create an issue to follow up this.

Comment (jeqo) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java]: https://issues.apache.org/jira/browse/KAFKA-10409

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java:191]: @ableegoldman I have extended `RocksDBTimestampedStoreTest` to use `reverseAll` and `reverseRange` as part of the current tests. 
Unfortunately, `AbstractKeyValueStoreTest` tests do not fit with the creation path of Timestamped stores as pre inserted data is required. 
Will add this to the same JIRA ticket to consider when refactoring iterators and tests.

Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java:191]: Ah, that makes sense I guess. Looks like the new additions to `RocksDBTimestampedStoreTest` cover the cross-column family code path so that's good enough for now

**Review by ableegoldman — COMMENTED** [review]
Thanks for the PR! Things LGTM now, I'll ping a committer to give this another pass and get it merged.

Let me know when the WindowStore PR is ready and I'll start reviewing the next one 😄

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java]: This warning seems to miss the most likely scenario, that the user just passed the arguments in the wrong order.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java:25]: Was this class unused or something?

**Review by vvcephei — COMMENTED** [review]
Thanks for the PR, @jeqo ! I made a partial pass today. It looks really good overall.

A high-level comment: in several places in this code, we have a internal boolean flag to represent the direction. It seems odd that in these places, `true = reverse`. It seems like the "natural" iteration order is forward, so we should have blocks like `if (forward) /*forward logic*/ else /*reverse logic*/`. But what we have here is `if (not reverse) /*forward logic*/ else /*reverse logic*/`. It's not exactly hard to understand, but still it seems to be creating unnecessary hoops for people to jump through when they are reading the code.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java:25]: yes, it looks like it was. Thanks for cleaning it up!

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbIterator.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbIterator.java:65]: clever!

**Review by vvcephei — COMMENTED** [review]
Hey @jeqo , I've completed my review. It looks good to me! Thanks so much for the KIP and PR.

Please let me know what you want to do about my prior comments, if anything :)

**Comment by jeqo** [comment]
@vvcephei thanks for your feedback! Inverting flag actually make things more readable. Changing to `forward` and `if (forward) {} else {}`.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java:25]: Yeah this was left over from the FKJ work IIRC

**Review by vvcephei — APPROVED** [review]
Thanks @jeqo !

**Comment by vvcephei** [comment]
It looks like java 14 only failed while attempting to download some docs from Oracle, and java 8 failed on this unrelated test: `org.apache.kafka.clients.admin.KafkaAdminClientTest.testMetadataRetries`

**Pull request merged**
